### ISSUE TITLE
Error squiggles with Unreal Engine 4.26 built from source on github and using latest patches

### ISSUE THREAD

**Issue by sean-mcmanus** [microsoft/vscode-cpptools#6358]
I am using latest Unreal Engine 4.26 built from source on github and using latest patches... VS code intellisense is half broken.. It shows error squiggles on freshly created classes.. It doesnt understand the  ::  or  \:    and suggests to use ;  instead in a freshly created c++ class.. 
![Screenshot from 2020-10-18 22-33-10](https://user-images.githubusercontent.com/40650341/96506818-e48f4400-1275-11eb-8a51-3c3d7a898eb7.png)

And this is a fresh created class of type actor. Vs code is not able to understand variables and this UWorldPosition::UWorldPosition() is also marked as error as ; 
As an officially supported IDE it always break with a certain extension or engine update :(

_Originally posted by @jatin-cbs in https://github.com/microsoft/vscode-cpptools/issues/5849#issuecomment-712420208_

**Comment by michelleangela** [comment]
@jatin-cbs

Could you provide your IntelliSense configuration settings by running the command `C/C++: Log Diagnostics`?

Could you also provide logs from enabling the [language server](https://code.visualstudio.com/docs/cpp/enable-logging-cpp#_enable-logging-for-the-language-server)? The logs will show which includePaths are parsed and sent to the IntelliSense engine. The first thing to ensure is that the required paths to any dependency libraries that the project needs is known to the IntelliSense engine.

**Comment by CreateLex** [comment]
There has been major bug for Mac users using Xcode or Visual Studio code. It still hasn't been resolved yet. I tried 4.26 previews and they still havent worked.

**Comment by michelleangela** [comment]
> There has been major bug for Mac users using Xcode or Visual Studio code. It still hasn't been resolved yet. I tried 4.26 previews and they still havent worked.

@CreateLex is there a reference (post/link) to an issue using XCode? Is this issue due to some change in Unreal Engine?

**Comment by CreateLex** [comment]
> > There has been major bug for Mac users using Xcode or Visual Studio code. It still hasn't been resolved yet. I tried 4.26 previews and they still havent worked.
> 
> @CreateLex is there a reference (post/link) to an issue using XCode? Is this issue due to some change in Unreal Engine?

Yes. this issue is due to an unreal engine generated project error. still waiting for them to resolve this issue for Mac users. using Xcode and visual studio code. The cpp defines array is empty.

You tube Link.

https://www.youtube.com/watch?v=oGUnRNMyjsU&t=5s

**Comment by Saladin1812** [comment]
@michelleangela  

https://pastebin.com/FefTGxj3

here is the log you asked for.

**Comment by michelleangela** [comment]
@jatin-cbs 

The logs from the language server look okay. Could you also send the IntelliSense configuration log by running `C/C++: Log Diagnostics` from the VS Code command palette?

The issue could be an incorrect IntelliSense configuration (like the file that has the class that is being used is not specified in the includePath property) _or_ there is an issue with Unreal engine.

**Comment by Saladin1812** [comment]
Its probably not an issue with the engine as codelite is working and also qtcreator with clang....intellisense which has been entirely reworked in 4.26 release as per the unreal engine github commits, is also having this issue and he is on Fedora 32 so it is also not distro specific.. maybe UBT bug

https://pastebin.com/1yjwHPgC  Here are the logs `C/C++: Log Diagnostics` @michelleangela

**Comment by boocs** [comment]
Check my guide for the #defines 4.26 bug. It might be the reason.
[https://gist.github.com/boocs/31e6c2e54815b7ba6246ee20315acf96#Step-10](https://gist.github.com/boocs/31e6c2e54815b7ba6246ee20315acf96#Step-10)

**Comment by Saladin1812** [comment]
@boocs so this is a unreal project generation bug in the UBT rather than the vscode extension I think @boocs  should this be closed ?

**Comment by boocs** [comment]
I'm not sure whose bug it is but the truth is it could be fixed rather easy with a UBT change. Should the first -include in a compile command file be blank? When researching the -include flag it did mention something special about the first -include command but I don't know.

I contacted Epic through a bug report about this twice. Once with a general problem and second time with the specific reason for the bug. I believe it was back in v. Preview 3 (we are now in v. Preview 5). They never added it to their bug tracker/issues though.

So:
This is a UE4 bug and Epic knows about this it and are going to fix it.
or
This isn't their bug and they're not going to patch something that isn't broke on their end.

**Comment by Saladin1812** [comment]
@boocs probably they won't in 4.25 there is a patch which someone sent as a pull request to fix the defines but it never got accepted and I don't think they will update unreal 4.25 branch any more after 4.25.4. Is compile commands in 4.26 faster parsing than the 4.25 tho..

**Comment by Saladin1812** [comment]
I will test it again and tell if this fixes it for me.. Thanks for the read me and you can also include https://gist.githubusercontent.com/ericwomer/142650e65473087073f30e5fb97fd6e8/raw/7955b060b4b8e7fcf2719630797a0de358092985/ue4_25-defines-fix.patch

This patch and apply it with...  git apply ue4_25-defines-fix.patch
Inside the unreal engine directory..
You can include this in your readme for permanent 4.25 intellisense fix..

**Comment by boocs** [comment]
It's a separate issue so probably not. It may help in Source files but mine has already been decently fast. At least once Intellisense loads on switching to a source file. Fire symbol on the bottom info bar is for context aware Intellisense and cylinder symbol for tag parser Intellisense.

When typing inside a function of a source file the fire symbol shouldn't come on unless you're creating a new function. The tag parser symbol may come on but doesn't seem to affect performance. If the fire symbol comes on when typing inside a function of a source file, and Intellisense is slow, then there's something wrong.

Header files are a different beast and are always slow. Whenever you type in a header file it recalculates Intellisense which can take 20 seconds.
So type 1 line... 20 secs.
Type another.... 20 secs.

I am working on extension that fixes this though by switching to Tag Parser Intellisense when editing a header file. Source files will still have Context Aware Intellisense because you'll definitely miss it with source files if it's gone.

**Comment by boocs** [comment]
> I will test it again and tell if this fixes it for me.. Thanks for the read me and you can also include https://gist.githubusercontent.com/ericwomer/142650e65473087073f30e5fb97fd6e8/raw/7955b060b4b8e7fcf2719630797a0de358092985/ue4_25-defines-fix.patch
> 
> This patch and apply it with... git apply ue4_25-defines-fix.patch
> Inside the unreal engine directory..
> You can include this in your readme for permanent 4.25 intellisense fix..

Does this need a full source code build to work?

**Comment by Saladin1812** [comment]
@boocs  This should probably work in the binary version as well.. you can apply it even if you downloaded Unreal Engine as zip or tar from github You just need git installed which probably you would have installed..  As the changes in this patch are done to 
/Engine/Source/Programs/UnrealBuildTool/Configuration/UEBuildModuleCPP.cs
These files are identical in the Binary version as well.. However I don't know how changes are applied in binary versions.. like make command..

**Comment by Saladin1812** [comment]
@sean-mcmanus   There is an issue now Unreal Engine 4.25 intellisense.. We fixed the defines and other stuff.. But now its showing error on include CoreMinimal.h 

cannot open source file "sys/sysctl.h" (dependency of "CoreMinimal.h")
CoreMinimal.h needs sys/sysctl.h    Is there something that needs to provide this can you confirm ?

**Comment by Saladin1812** [comment]
Cannot open source file "sys/sysctl.h"  means it is expecting to find this file somewhere.. Can someone investigate this..

**Comment by Saladin1812** [comment]
Also if I can skip this sys/sysctl.h would be great..  Because of 1 wrong error in Unreal Include paths Vscode disables error checking in the whole cpp file

**Comment by Saladin1812** [comment]
Oh I got the issue 
/home/shuriken/Applications/UnrealEngine/Engine/Extras/ThirdPartyNotUE/SDKs/HostLinux/Linux_x64/v16_clang-9.0.1-centos7/x86_64-unknown-linux-gnu/usr/include/

This is the path I added to includePaths and it solved the issue.. The main issue is Unreal Engine uses bundled clang and VS Code uses system clang. So including this solves it as sys/sysctl.h is a part of this directory :) Finally everything is working and intellisense is  accurate and fast Thanks for everyone's help @sean-mcmanus @boocs  

:+1:

**Comment by michelleangela** [comment]
@jatin-cbs If UnrealEngine ships with its own compiler, you can set the property "compilerPath" to that path as the C++ extension tries to get includes based off the compilerPath.

**Comment by boocs** [comment]
I believe I have the same problem just came to the wrong conclusion with my 4.26 bug guide even though the fix works for me.

 The first command should be the compiler path and UE4 doesn't set it for my config (Windows) and I guess other configs as well. So I'll have to write another UE4 bug report, with the correct info, if someone hasn't done so already.

VSCode does now complain about this as well with a bunch of errors like so:
Attempting to get defaults from C++ compiler in compile_commands.json file: '-include'
Specified compiler was not found: -include

I don't remember it doing so before.

**Comment by Saladin1812** [comment]
@boocs I don't think compiler path is the main issue as on Linux Unreal Engine uses clang for compilation and if clang is not found or not mentioned in the json file it defaults to  usr/bin/clang
So not finding a compiler is not an issue on Linux if you have clang installed at a system level. So my compiler is working but still those same errors on 4.26.. your fix did worked tho.

**Comment by boocs** [comment]
I updated my fix. Has nothing to do with the -include flag. Just need some placeholder as first command. Let me know if this fix works as well.

```
"command": "
"command": "\"\"
```

https://gist.github.com/boocs/31e6c2e54815b7ba6246ee20315acf96#Step-10

**Comment by boocs** [comment]
I created a barebones, nothing to do with Unreal, test and it acts the same way.

So it could be Epics 4.26 bug - They should put a \"\" in place of a compiler path and let VSCode fallback to a compiler it finds.

It could also be VSCode's extension bug - They shouldn't throw away the first parsed command if it's determined to not be a compiler. But then you'd have to determine if it's a bad compiler path or an actual compiler flag.

It might be worth asking VSCode extension makers to fix this:
If the first command, in a compilation database, starts with a '-' then don't throw it away. Try to find a fallback compiler to use and use every compiler flag after "command": when found.

**Comment by Saladin1812** [comment]
@boocs  Its working.. yes an extension fix may be good @sean-mcmanus  any suggestions regarding this ? We will have to replace all

`"command": "`
TO
`"command": "\"\"`
in compileCommands_YourProjectName.json

for 4.26 unreal release.. 
Can you make sure this is an extension bug or an engine upstream bug ?

**Comment by sean-mcmanus** [comment]
Sure, I can create a bug -- but does someone have a simple repro?

**Comment by boocs** [comment]
I can put up the one I tested before. I'll need to look at it again to make sure everything is good and I'll post the link.

**Comment by boocs** [comment]
Here it is: [https://github.com/boocs/compilecommands](https://github.com/boocs/compilecommands)

**Comment by boocs** [comment]
4.26.0 was just released (Out of Preview). No fixes.

The workaround I posted still works.

**Comment by Saladin1812** [comment]
@sean-mcmanus  A simple repro is 

1. Install Unreal Engine 4.26 using the Epic Games Launcher or built it from source on Windows or Linux respectively.
2. Having VS Code with latest cpp extension.. Create a new cpp project, Generate the vscode project files

Expected result:  After all the parsing and updating intellisense is done the vscode intellisense should understand classes :: and not show other false errors or incomplete intellisense

FIX :   replacing all 
`"command": "` 
TO
`"command": "\"\"`
in compileCommands_YourProjectName.json file.

Here is the sample project created by @boocs which explains the error even further  https://github.com/boocs/compilecommands

**Comment by sean-mcmanus** [comment]
Okay, thanks for the repro info.

@michelleangela Were you planning to look at this? If not, you can re-assign it to me.

**Comment by jatin-cbs** [comment]
@sean-mcmanus 
After my discussion with one of the teaching assistants at gamedev.tv they made a video explaining this error in Vscode. Please take a look at:

https://www.gamedev.tv/courses/learn-unreal-engine-c-developer-4-22-for-video-game-development/lectures/17467871

@boocs  Can I have access of https://gist.github.com/boocs/31e6c2e54815b7ba6246ee20315acf96 to submit you some pull requests regarding Linux specific fixes for VScode ? Won't be more than a couple of lines.. Thanks.

**Comment by boocs** [comment]
> Can I have access of https://gist.github.com/boocs/31e6c2e54815b7ba6246ee20315acf96 to submit you some pull requests regarding Linux specific fixes for VScode ? Won't be more than a couple of lines.. Thanks.

@jatin-cbs Sent invite

**Comment by sean-mcmanus** [comment]
According to the docs at https://clang.llvm.org/docs/JSONCompilationDatabase.html it's not a valid case. Did someone file a bug on Unreal Engine or did they explain why they're generating the compile commands like that?

Do you think we should add the workaround or get Unreal to fix it?

**Comment by boocs** [comment]
> Did someone file a bug on Unreal Engine or did they explain why they're generating the compile commands like that?
> 
> Do you think we should add the workaround or get Unreal to fix it?


I filed a report but it never got posted. 

Also they probably shouldn't be using compile commands. If you look at the compile commands file they generate, all files have the same command args. If that's the case, they should go back to generic way of doing Intellisense before 4.26.

**Comment by CreateLex** [comment]
Filed a report for for Unreal Engine since 4.25 and they have still had no fixed to this problem. it might be a very difficult but because they still have not fixed it. The problem also affects Xcode as well.

**Comment by sean-mcmanus** [comment]
That's fine -- our pending 1.2.0-insiders should have a workaround for it.

**Comment by CreateLex** [comment]
@sean-mcmanus  Thank you for the update. Waiting on Unreal Engine or vscode work around.

**Comment by boocs** [comment]
For my extension, I'm creating a better Backup Intellisense that parses the compile commands for includePaths which also lets you convert it to exclusive includePath Intellisense.

Doing so shows 88 include paths, in the compile commands file, that were incorrect. They must be not that important since Intellisense still worked.

Would this be considered a bug(or feature request) that the cpptools extension doesn't warn about incorrect include paths in the compile command's file?

**Comment by sean-mcmanus** [comment]
@boocs I think it would be a matter of opinion. I would call it an "enhancement". I could file an issue for you if you want or you could file one.

**Comment by boocs** [comment]
@sean-mcmanus  Sure go ahead 👍

**Comment by boocs** [comment]
I hope you don't mind but something else I always wondered about. I'm trying to find the best configuration for multiple workspaces. Do you know why this is?

Compile commands:
Include Paths : (Both workspaces with all paths in browse.path) :
Include Paths: (UE4 workspace has all UE4 paths in browse.path other workspace has rest of the paths):

All three have a tag parser size of around 1.6 GB

If I do it this way:
Main Workspace has all paths in browse.path and UE4 workspace browse.path has an empty array

Tag parser size is less than 1 GB. If I do this with UE4 4.25 it reduced it by more than 1/2.

Switching between source/header, symbols, go to definition seem unaffected by this in both workspaces.

**Comment by sean-mcmanus** [comment]
I don't fully understand what you're describing, but I'll just explain how it works. The code will iterate over each workspace folder one at a time and apply that workspace folder's settings (files.exclude, limitSymbolsToIncludedHeaders, etc.) in regards to which files will get tag parsed, but all the symbols get stored to a single database that is shared by all the workspace folders, which enables symbols across all the workspace folders to be located. With 0.26.3 or older, every workspace folder had it's own database and symbol operations didn't work between them. Does that answer your question?

**Comment by boocs** [comment]
I updated to make it more clear.

I'll just use the empty browse.path array method. Tag parser functions seem a lot snappier with seemingly no loss of functionality.

**Comment by sean-mcmanus** [comment]
Yeah, as I stated in the other posts, the behavior you're seeing is not expected and I'm not aware of a repro, but there shouldn't be any downside, other than any inconvenience when adding a new folder.

**Comment by sean-mcmanus** [comment]
We added a workaround to handle the missing compilerPath with https://github.com/microsoft/vscode-cpptools/releases/tag/1.2.0-insiders2 -- also, Unreal Engine may be updating their compile_commands.json with a valid compiler...it's possible we could revert our workaround at some point, since the compile_commands without a compiler path doesn't appear to be valid.

**Comment by boocs** [comment]
> We added a workaround to handle the missing compilerPath with https://github.com/microsoft/vscode-cpptools/releases/tag/1.2.0-insiders2 -- also, Unreal Engine may be updating their compile_commands.json with a valid compiler...it's possible we could revert our workaround at some point, since the compile_commands without a compiler path doesn't appear to be valid.

Good to hear!

**Comment by jatin-cbs** [comment]
> We added a workaround to handle the missing compilerPath with https://github.com/microsoft/vscode-cpptools/releases/tag/1.2.0-insiders2 -- also, Unreal Engine may be updating their compile_commands.json with a valid compiler...it's possible we could revert our workaround at some point, since the compile_commands without a compiler path doesn't appear to be valid.

Yes there was an upstream commit recently 
https://github.com/EpicGames/UnrealEngine/commit/c3785db9c96ac0790df09a6b184f2b1d7e8b6802
For VS Code.. Let me test this.

**Comment by Krzysow** [comment]
I've updated the C/C++ extension to 1.2.0 as well as UE to 4.26.1 but I'm still getting error squiggles throughout.
<img width="1680" alt="Screenshot 2021-02-16 at 18 57 36" src="https://user-images.githubusercontent.com/48242451/108102634-15974e00-7089-11eb-9f9c-d292e0fbc8f5.png">

**Comment by Krzysow** [comment]
I made a new project as well, but the same thing happens.

**Comment by CreateLex** [comment]
This issue is still present on all MacOS. But appears to be fixed on windows.  :(

**Comment by boocs** [comment]
On  your new 4.26.1 project, if you open your main workspace's c_cpp_properties.json what is your intelliSenseMode set to? 

Is your compileCommands set? If you compileCommands is set and you open the file path it is pointing to what is the "command" variable set to? Is it a huge list of directories or does it point to a response file?

If it points to a response file, do all files in the compile commands file have the same response file set in the "command" var?

**Comment by Krzysow** [comment]
@boocs
1. c_cpp_properties - "intelliSenseMode": "clang-x64".
2. the compileCommands is set and the "command" variable points to a rsp file.
3. In my new project there are 5 entries and they all have the same "command" set.
<img width="1680" alt="Screenshot 2021-02-17 at 08 55 59" src="https://user-images.githubusercontent.com/48242451/108173135-0bfbfd80-70fe-11eb-9fb5-16357b2779a3.png">
<img width="1680" alt="Screenshot 2021-02-17 at 08 56 04" src="https://user-images.githubusercontent.com/48242451/108173154-10281b00-70fe-11eb-85f1-aa9b8aa39388.png">

**Comment by SuperWig** [comment]
Is the fact that Unreal hardcodes cl.exe at least part of the issue?

**Comment by Krzysow** [comment]
Probably. When I find and replace every "cl.exe" with "clang" the errors go away but the intellisense still doesn't work. It only autocompletes pure c++ stuff, not even variable names.

**Comment by boocs** [comment]
First change your cppStandard to "c++14" instead of "c++98". Did you set that or did UE4 set that for you? Seems weird if UE4 set that.

UE4 uses C++14 by default and if you get the full source code you can compile it to C++17.

Hopefully it's that simple. With my Windows version, VSCode will automatically switch to C++14 if I try to set my UE4 project to "c++98". Maybe something is different with the Mac version when using cl.exe when it's not suppose to.

Even though the response file contains command line options for cl.exe, the command line options should be compatible with clang. Your pic does show VSCode changing the compile to clang so it should work.

Maybe your cpp extension logs will say something. Do steps 1-5:
https://code.visualstudio.com/docs/cpp/enable-logging-cpp#_enable-logging-for-the-language-server

**Comment by sean-mcmanus** [comment]
Let us know you think we have a bug with any new compile commands that Unreal is generating.

**Comment by Krzysow** [comment]
The cppStandard was generated by Unreal. After changing it to c++14 I refreshed the project from engine and it went back to c++98.

But the change worked. The intellisense is back and it was a little slow at first but I think it's fine now.

Also here is the cpp extension log: https://pastebin.com/6ePkMeHR

**Comment by boocs** [comment]
On Windows when you add a class the compile commands file is automatically updated. Maybe it's not on Mac. Try resetting your project again. You'll then have to change the cppStandard again.

**Comment by Krzysow** [comment]
Yeah, that was the case. The compileCommands didn't have the 2 new paths until I refreshed the project. I guess I'll have to refresh the project and change the cppStandard every time I add new files for now.

**Comment by boocs** [comment]
You can try to report a bug in the help menu of the engine about these two issues.

I'll add the cppStandard fix to my VSCode extension and see about compile command creation so you don't have to reset every time. My extension currently doesn't work with 4.26.1. I didn't know typescript much so it was hastily written to get it to work with 4.25 and 4.26.0. I'm now currently converting it over to classes with more 'proper' code to enable 4.26.1+ support. It'll be awhile unfortunately...

**Comment by jatin-cbs** [comment]
Don't open the issue regarding intellisense fix there are already 2 vscode project file generation changes in the upstream after the 4.26.1 release which are working atleast on Linux. Most windows users don't have GitHub build so I don't know how can you test the latest patches but it should fix it on windows as well. You may ask in forums or answers website.

**Comment by jatin-cbs** [comment]
By not reporting an issue I meant in the Unreal Engine issue tracker, as they can't backport changes to the already released branch, so you will have to wait for 4.26.2

**Comment by jatin-cbs** [comment]
Anyone who has Unreal Engine source code access, check this commit https://github.com/EpicGames/UnrealEngine/commit/44a4fcf03b56751709e8ce87c8c43e9f25b26a43

**Comment by jatin-cbs** [comment]
Now All Windows Mac OS and Linux will have same default configuration for better intellisense, more specific one can be manually selected by the user @sean-mcmanus  @boocs  Can you check this commit ?

**Comment by boocs** [comment]
@jatin-cbs 
Don't see anything obvious that would fix the Mac bug so hopefully it's there.

Also based on a comment they're gonna leave cl.exe as default and just let VSCode switch to clang when needed.

**Comment by Krzysow** [comment]
I don't mind changing the cppStandard once in a while until either the extension or the 4.26.2 comes out so I'm not going to open a new issue for now as suggested.

**Comment by boocs** [comment]
@Krzysow

It's actually VSCode that is setting the cppStandard incorrectly for UE4. You just need to set the cppStandard in the global User settings so VSCode doesn't try to guess the cppStandard of the UE4 project.

Tested this on Windows. I already had my global User cppStandard set. When I removed it, VSCode incorrectly sets the project to C++17.

**Comment by Krzysow** [comment]
@boocs

When I set the default cppStandard in VSCode to c++14 it's not showing up in the c_cpp_properties.json file but there are no errors like before so I assume it's set somewhere internally. The problem with compileCommands file not updating after adding new stuff is still there but at least it's one less step I have to do each time I add a new file.

Thanks for the solution.

**Comment by boocs** [comment]
@Krzysow 
My extension should be released today. I added the feature to automatically add missing source/header files to compile commands. I wont support 4.26.0 but will support 4.26.1+.

I was about to release it yesterday until I remembered the cppStandard issue. Oops.

**Comment by Krzysow** [comment]
@boocs 
Great, I'll definitely check for the extension in the evening.
